### PR TITLE
Support for beta6

### PR DESCRIPTION
**Pull request by gfrappier**
Modified configuration and code to support beta6

### CODE REVIEW

File: vNext/src/Microsoft.AspNet.OData/DefaultODataModelProvider.cs [hunk 1]
```diff
@@ -22,6 +22,34 @@ public static IEdmModel BuildEdmModel(Type ApiContextType)
                 builder.AddEntitySet(property.Name, entity);
             }
 
+            // Get the actions and functions into the model
+            var publicMethods = ApiContextType.GetMethods(BindingFlags.Public | BindingFlags.Instance);
```
Comment (congysu) [inline vNext/src/Microsoft.AspNet.OData/DefaultODataModelProvider.cs]: Update or add a sample about actions and functions.

File: vNext/src/Microsoft.AspNet.OData/DefaultODataModelProvider.cs [hunk 2]
```diff
@@ -22,6 +22,34 @@ public static IEdmModel BuildEdmModel(Type ApiContextType)
...
+            foreach (var method in publicMethods)
+            {
+                if (!method.IsSpecialName)
+                {
+                    var entityClrType = TypeHelper.GetImplementedIEnumerableType(method.ReturnType) ?? method.ReturnType;
+                    ProcedureConfiguration configuration;
+
+                    if (entityClrType.Name != "Void")
```
Comment (congysu) [inline vNext/src/Microsoft.AspNet.OData/DefaultODataModelProvider.cs]: entityClrType != typeof(void)

File: vNext/src/Microsoft.AspNet.OData/DefaultODataModelProvider.cs [hunk 3]
```diff
@@ -22,6 +22,34 @@ public static IEdmModel BuildEdmModel(Type ApiContextType)
...
+                    {
+                        var returnType = builder.AddEntityType(entityClrType);
+                        configuration = builder.Function(method.Name);
+                        configuration.ReturnType = returnType;
+                    }
+                    else
+                    {
+                        configuration = builder.Action(method.Name);
```
Comment (congysu) [inline vNext/src/Microsoft.AspNet.OData/DefaultODataModelProvider.cs]: Action can also have a return type, we shall define whether an operation is side effecting to differentiate between action and function.

File: vNext/src/Microsoft.AspNet.OData/EdmLibHelpers.cs [hunk 4]
```diff
@@ -188,7 +188,20 @@ public static IEdmTypeReference GetEdmTypeReference(this IEdmModel edmModel, Typ
                 Type elementType;
                 if (clrType.IsCollection(out elementType))
                 {
-                    edmType = new EdmCollectionType(edmModel.GetEdmType(elementType).ToEdmTypeReference(IsNullable(elementType)));
```
Comment (congysu) [inline vNext/src/Microsoft.AspNet.OData/EdmLibHelpers.cs]: Do we encounter any failure here?

File: vNext/src/Microsoft.AspNet.OData/Query/ODataQueryOptions.cs [hunk 5]
```diff
@@ -47,7 +47,10 @@ public ODataQueryOptions(ODataQueryContext context, HttpRequest request)
                 context.ElementType,
                 context.NavigationSource,
                 queryOptionDict);
-            
+
+            // Support for $expand
+            request.ODataProperties().SelectExpandClause = this._queryOptionParser.ParseSelectAndExpand();
```
Comment (congysu) [inline vNext/src/Microsoft.AspNet.OData/Query/ODataQueryOptions.cs]: Should this be in `BuildQueryOptions`, similar to $filter.

File: vNext/src/Microsoft.AspNet.OData/Routing/Conventions/DefaultODataRoutingConvention.cs [hunk 6]
```diff
@@ -59,6 +59,7 @@ public ActionDescriptor SelectAction(RouteContext routeContext)
                 if (structuralPropertySegment != null)
                 {
                     routeTemplate += "/" + structuralPropertySegment.Property.Name;
+                    methodName += structuralPropertySegment.Property.Name;
```
Comment (congysu) [inline vNext/src/Microsoft.AspNet.OData/Routing/Conventions/DefaultODataRoutingConvention.cs]: `methodName` is Post, Get, Patch, Put, or Delete.

File: vNext/src/Microsoft.AspNet.OData/Routing/Conventions/DefaultODataRoutingConvention.cs [hunk 7]
```diff
@@ -80,14 +81,61 @@ public ActionDescriptor SelectAction(RouteContext routeContext)
             
             var services = routeContext.HttpContext.ApplicationServices;
             var provider = services.GetRequiredService<IActionDescriptorsCollectionProvider>();
-            var actionDescriptor = provider.ActionDescriptors.Items.SingleOrDefault(d =>
+
+            var methodDescriptor = new List<ActionDescriptor>();
```
Comment (congysu) [inline vNext/src/Microsoft.AspNet.OData/Routing/Conventions/DefaultODataRoutingConvention.cs]: methodDescriptors

File: vNext/src/Microsoft.AspNet.OData/Routing/Conventions/DefaultODataRoutingConvention.cs [hunk 8]
```diff
@@ -80,14 +81,61 @@ public ActionDescriptor SelectAction(RouteContext routeContext)
...
+            foreach (var descriptor in methodDescriptor)
             {
-                var c = d as ControllerActionDescriptor;
-                return c != null && c.ControllerName == controllerName &&
-                    (controllerName == "Metadata" ||
-                        ((HttpMethodConstraint)c.ActionConstraints.First()).HttpMethods.Contains(methodName) &&
-                         c.AttributeRouteInfo.Template.EndsWith(routeTemplate));
-            });
+                bool matchFound = true;
+                if (descriptor.Parameters.Count(d => d.BindingInfo == null) == keys.Count)
```
Comment (congysu) [inline vNext/src/Microsoft.AspNet.OData/Routing/Conventions/DefaultODataRoutingConvention.cs]: How about action and function - `OperationSegment`

**Comment by lewischeng-ms** [comment]
@gfrappier Thanks for your contribution. The change for supporting beta6 has been merged as https://github.com/OData/WebApi/commit/f4fb5d9d3795a355c74624ab7802a26e717fa933. The other change will be merged separately.

**Pull request closed without merging**